esmvalcore/_data_finder.py

```diff
@@ -241,11 +241,21 @@ def _find_input_files(variable, rootpath, drs):
 
 def get_input_filelist(variable, rootpath, drs):
     """Return the full path to input files."""
-    # change ensemble to fixed r0i0p0 for fx variables
-    # this is needed and is not a duplicate effort
-    if variable['project'] == 'CMIP5' and variable['frequency'] == 'fx':
-        variable['ensemble'] = 'r0i0p0'
     (files, dirnames, filenames) = _find_input_files(variable, rootpath, drs)
+    if (
+        variable['frequency'] == 'fx'
+        and variable['ensemble'] != 'r0i0p0'
+        and dirnames == []
```

Comment:
you are strangling the code here ie you are looking for a very specific combination of parameters that may result in the combined condition to be missed a lot of the times. Note that for CMIP5 the ensemble is always `r0i0p0` for fx variables, and the ones you found out that don't stick to this rule should still have the fx data files in `/r0i0p0` dir - if not, that's a problem of the data release. For CMIP6 they will never be in `r0i0p0` but dotted around in various ensembles `rXiXpXfX`, as you well note, but not always in *all* ensemble dirs (they may be in `r1i1p1f1` but not in `r2i1p1f1`) so a `*` search on the ensemble dir will do the job, or a quasi-explicit `r*` for ensemble, the fx data files should all be the same no matter the ensemble. So CMIP6 should be covered as it is now, the only problem I see is the CMIP5 ones that are not `r0i0p0`, which kind of alarms me and am thinking what sort of files are those

Reply:
Thanks for your comments. I'm not sure how a user would write a recipe to handle data where the ensemble containing the fx info is stored under a different ensemble to that specified in the variable (tas, pr or whatever), which as you say can happen under CMIP6.
Similarly for CORDEX, the data specification I linked to above states that the fx info can live either under the r0i0p0 ensemble or the ensemble of the regular variables in question.

My issue is I'm trying to develop a recipe that performs the same analysis on large multi model ensembles from different projects (~30 models each from CMIP5/6, CORDEX and UKCP) I want to include a pre processor that does the land sea masking of the models..
i.e.
```
    mask_landsea:
      mask_out: sea
```
In the current version of the code the data finder looks in `r0i0p0` for CMIP5, and the same ensemble as the variable for everything else. So at the moment this will not find the fx files for a large proportion of the CMIP6 and CORDEX data. I don't see any way in a recipe, short of writing a separate pre-processor for each dataset, to deal with looking in different ensemble folders depending on the dataset.
One other issue I've also just discovered, is for some of the CORDEX datasets, sometimes the data exists under the historical experiment, but not the rcp85 for example. :-/ As you say, the fx files should be the same across ensembles and experiments, so ideally it would be good to search across.

Are you suggesting a modification of some sort to the https://github.com/ESMValGroup/ESMValCore/blob/574f9f8054a99aedfc4b4606682f367ad82b1e6d/esmvalcore/_data_finder.py#L201 function to replace the ensemble part of the dir name with a wildcard?

Reply:
cheers for the detailed explanation :beer: - yes, probably a wildcard would be better, or better off, a the use of `glob.glob()` with a wildcard. Do you think that'd be beyond the scope of this PR?